module_utils/foreman_helper.py

```diff
@@ -51,22 +53,189 @@ def filter_module_params(self):
+                            new_value = type(value)(new_value)
+                    if isinstance(value, list) and value and isinstance(value[0], dict) and 'id' in value[0]:
+                        value = [{'id': item['id']} for item in value]
+                    elif isinstance(value, dict) and 'id' in value:
+                        value = value['id']
+                    if isinstance(new_value, dict) and 'id' in new_value:
+                        new_value = new_value['id']
+                    if not value == new_value:
+                        volatile_entity[key] = new_value
+                        fields.append(key)
+                # check_missing is a special case, Foreman sometimes returns different values
+                # depending on what 'type' of same object you are requesting. Content View
+                # Filters are a prime example. We list these attributes in `check_missing`
+                # so we can ensure the entity is as the user specified.
+                if check_missing is not None and key not in entity_dict and key in check_missing:
+                    volatile_entity[key] = None
+                    fields.append(key)
+            if force_update:
+                for key in force_update:
+                    fields.append(key)
+            if check_missing is not None:
+                for key in check_missing:
+                    if key in entity_dict and key not in volatile_entity.keys():
+                        volatile_entity[key] = new_value
+                        fields.append(key)
+            if len(fields) > 0:
+                if not self.check_mode:
+                    new_data = {}
+                    for key, value in volatile_entity.items():
+                        if key in fields:
+                            new_data[key] = value
+                    result = self.foremanapi.resource(resource).call('update', {'id': entity_id, hash_entry_name: new_data})
+                return True, result
+            return False, result
+        except Exception as e:
+            self.fail_json(msg='Error while updating {0}: {1}'.format(
+                resource, str(e)))
+
+    def ensure_resource_state(self, resource, entity_dict, entity, state, check_missing=None, check_type=None, force_update=None):
+        changed, _ = self.ensure_resource(resource, entity_dict, entity, state, check_missing, check_type, force_update)
+        return changed
+
+    def ensure_resource(self, resource, entity_dict, entity, state, check_missing=None, check_type=None, force_update=None):
+        """ Ensure that a given entity has a certain state """
+        changed, changed_entity = False, entity
+
+        # FIXME: this is a hack I want to get rid of before merging
+        # the "issue" is that the resource in foreman is called "domains",
+        # but when you post data to it, the entry has to be called "domain"
+        # see https://theforeman.org/api/1.21/index.html
```

Comment:
is this in translation in apidoc? I dont think it is...


Nailgun uses inflection, but I'd rather avoid another dep and inflection isn't currently packaged as RPM :-1:  https://github.com/SatelliteQE/nailgun/blob/fe1593adcabe89c46796e95df51322171f11915b/nailgun/entity_mixins.py#L278 

Sorry, i have no good solution

Reply:
Yeah, I looked at `inflection` and found it way too big for that small part we'd be actually using.

And no, it's not in the apidoc, unless I go and iterate over the possible params and "guess" (uh)